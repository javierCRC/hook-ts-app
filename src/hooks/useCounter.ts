
import React, { useState } from 'react'

export const useCounter = ( pValue: number = 105) => {
    
    const [state, setState] = useState(pValue);

    const f_Increment = (pFactor: number) => {
        setState ( state + pFactor );
    }

    const f_Reset = () => {
        setState ( pValue );
    }

    const f_Decrement = (pFactor: number) => {
        setState ( state - pFactor );
    }

    return {
        state,
        f_Increment,
        f_Decrement,
        f_Reset
    };

}
