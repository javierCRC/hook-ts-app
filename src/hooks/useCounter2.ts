
import  {  useState } from 'react'

export const useCounter2 = ( pValue: number = 1) => {
    
    const [counter, setCounter] = useState(pValue);

    const f_Increment = () => {
        setCounter ( counter + 1 );
    }

    const f_Reset = () => {
        setCounter ( pValue );
    }

    const f_Decrement = () => {
        setCounter ( counter - 1 );
    }

    return {
        counter,
        f_Increment,
        f_Decrement,
        f_Reset
    };

}
