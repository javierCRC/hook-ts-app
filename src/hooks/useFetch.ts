import { useEffect, useRef, useState } from 'react'

export const useFetch = ( pUrlEndPoint: string , pInitialRequest: any ) => {
   
    // example of use a hook with real example
    const vIsComponentMounted = useRef( true );
    
    const [state, setState] = useState(pInitialRequest);
    
    //call the operation one time when the component is loaded.
    useEffect(() => {
        
        return () => {
            // the component is dismounted
            vIsComponentMounted.current = false;
        }
    },[])

    // only change when the pUrlEndPoint changed.
    useEffect(() => {

        setState(pInitialRequest);
        
        fetch (pUrlEndPoint)
          .then ( respo => respo.json())
          .then( data => {

              //setTimeout(() => {

                if( vIsComponentMounted.current ){
                    setState({
                        pLoading:false,
                        pError: null,
                        pData: data
                    });
                }
                /*else{
                    alert("The component is dismounted");
                } */
                
            //}, 2000);

          }).catch(()=> {
            setState({
                pLoading:false,
                pError: 'dont be obtain the data',
                pData: null
            });
      });

    }, [pUrlEndPoint])
    
    return state;

}
