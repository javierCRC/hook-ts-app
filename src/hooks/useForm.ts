import { useState } from "react";
//import { IFormNew } from "../components/interfaces";


export const useForm = ( pObject: any = {} ) => {
     

    const [stateValues, setStatesValues] = useState(pObject);
    
    const f_handleInputChange = ( { target }: any ): any => {
        //console.log(target);
        
        setStatesValues( {
           ...stateValues,
           [ target.name] : target.value
         }
        );
   }

   const f_handleCleanForm = () => {
    setStatesValues(pObject);
 }

   //return [ stateValues , f_handleInputChange , f_handleCleanForm ];
   return { stateValues , f_handleInputChange , f_handleCleanForm };
}
