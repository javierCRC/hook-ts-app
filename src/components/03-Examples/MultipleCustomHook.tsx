import React from 'react'


import { useFetch } from '../../hooks/useFetch';
import { useCounter2 } from '../../hooks/useCounter2';

import './effects.css';
import { IReqQuote } from '../interfaces';


export const MultipleCustomHook = () => {

    const { counter , f_Increment, f_Decrement} = useCounter2(1);
    
    const objRequest: IReqQuote = {
        pData: [{
            quote_id:0,
            quote:'',
            author:'',
            series:''
        }],
        pLoading: true,
        pError: ''
    }
    
    const vURLAPI= `https://breakingbadapi.com/api/quotes/${counter}`;
    const state = useFetch(vURLAPI,objRequest);
    
    //console.log(state);
    
     const { pLoading, pData } = state;
     
     console.log(pData);

     const { quote , author  } = !!pData && pData[0]; // if exist data and is not null , extract the data of the position 0 of the array.

    return (
        <div>
            <h1> Breaking bad quotes TypeScript</h1>
            <hr />

             {
                 // ternaring condition
                 pLoading
                 ? // if
                    (
                        <div className='alert alert-info text-center'> 
                            Loading...
                        </div>
                    )
                 : // else
                    (
                        <blockquote className='blockquote text-right'>
                        <p className='mb-0'> { quote } </p>
                        <footer className='blockquote-footer'> { author } </footer>
                        </blockquote>
                     )
             }
            

        <button className="btn btn-primary m-2"
                onClick={ f_Decrement }>
                prev quote
        </button>

        <button className="btn btn-primary"
                onClick={ f_Increment }>
                Next quote
        </button>
            

        </div>
    )
}

