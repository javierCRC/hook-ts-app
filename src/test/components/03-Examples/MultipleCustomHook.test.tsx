import { renderHook,act } from '@testing-library/react-hooks';
import { MultipleCustomHook } from '../../../components/03-Examples/MultipleCustomHook';
import { shallow } from "enzyme"
import { useFetch } from '../../../hooks/useFetch';
import { useCounter2 } from '../../../hooks/useCounter2';
//import '@testing-library/jest-dom';


jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter2');


describe('Test in component <MultipleCustomHook />', () => {

    
    
    test('1.Must do render the component MultipleCustomHook.', () => {
        
        
        // 1 Arrange *******************************************************************************
           
             // @ts-ignore
            useCounter2.mockReturnValue({
                counter: 10,
                f_Increment: ()=> {},
                f_Decrement: ()=> {}
            }); 

         // @ts-ignore
           useFetch.mockReturnValue({
              pData: null,
              pLoading: true,
              pError: null
           });

        //2. Act *******************************************************************************

             const cWrapper = shallow(<MultipleCustomHook />);
    
        //3. Assert *******************************************************************************

            expect(cWrapper).toMatchSnapshot();
            
    });

    test('2.should do show the information about the component.', () => {
        
        
        // 1 Arrange *******************************************************************************
                
                // @ts-ignore
                useCounter2.mockReturnValue({
                    counter: 10,
                    f_Increment: ()=> {},
                    f_Decrement: ()=> {}
                });

                // @ts-ignore
                useFetch.mockReturnValue({
                    pData: [{
                        author: 'Javier Mendoza Blandon',
                        quote: " The good engeniering solve."
                    }],
                    pLoading: false,
                    pError: null
                });

                let vAlertLoading: boolean;
                let vQuoted: string;
                let vAuthor: string;

        //2. Act *******************************************************************************

            const cWrapper = shallow(<MultipleCustomHook />);
                  
                   //console.log(cWrapper.html() );

                vAlertLoading = cWrapper.find('.alert').exists();
                vQuoted = cWrapper.find('.mb-0').text().trim();
                vAuthor = cWrapper.find('footer').text().trim();
                  //console.log('2.....Alert: '+vAlertLoading+' -> Quote: '+vQuoted+' -> Author: '+vAuthor);
        
        //3. Assert *******************************************************************************

            expect(vAlertLoading).toBe(false);
            expect(vAuthor).toBe('Javier Mendoza Blandon');
            expect(vQuoted).toBe('The good engeniering solve.');      
    });

    
});
