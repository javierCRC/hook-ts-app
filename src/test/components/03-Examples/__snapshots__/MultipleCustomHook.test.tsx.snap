// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in component <MultipleCustomHook /> 1.Must do render the component MultipleCustomHook. 1`] = `
<div>
  <h1>
     Breaking bad quotes TypeScript
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary m-2"
    onClick={[Function]}
  >
    prev quote
  </button>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;
